File: python/mcts.py
import time
from board import Board
import copy

class MCTS(object):

    def __init__(self, state, player, time=6):
        self.root = Board(state, player, None)
        self.root.action = None
        self.player = player
        self.time = float(5)
        self.N , self.Q, self.C = 0.001, 0, 1.95


    def search(self):

        curState = self.root
        if(len(self.root.available) == 0):
            print("No choice!")
            return self.root
        cor_avail = curState.corner()
        if(cor_avail is not None):
            return self.directBoard(curState, cor_avail)
        beg = time.time()
        while(True):
            curState = self.root
            for i in range(5):
                next_board = self.treePolicy()
                reward = self.defaultPolicy(curState)
                next_board.backup(reward, curState)
            if((time.time() - beg) > self.time):
                break
        return self.root.bestChild(self.C)



    def directBoard(self, curState, cor_avail):

        ret = copy.copy(self.root)
        ret.action = cor_avail
        ret.state[cor_avail[0]][cor_avail[1]] = ret.player
        return ret
        
            

    def treePolicy(self):

        curState = self.root
        while(True):
            if(curState.notEnd()):
                break
            if(curState.needExpand()):
                return curState.expand()
            else:
                if(len(curState.available) == 0):
                    break
                curState = curState.bestChild(self.C)
        return curState
    

    def defaultPolicy(self, curState):
        
        if(len(curState.available) == 0):
            nextState = copy.copy(curState)
            nextState.player = -nextState.player
            return defaultPolicy(nextState)
            
        while(curState.notEnd()):
            curState = curState.randomGo()
  
        return curState.isWinner()      
         

        




        
         
        

File: C++/MCTS.cpp
#include<iostream>
#include<time.h>
#include"MCTS.h"



MCTS::MCTS(double p[4], char state[8][8], int _player, double _C)
{
	player = _player;
	C = _C;
	root = new Board(p, state, player, NULL, 2);
}

Board* MCTS::search()
{
	Board* nextBoard;
	Board* curState = root;
	int i;
	double reward;
	time_t beg;
	if (root->numAvailable == 0)
	{
		return NULL;
	}

	int k = 1;
	if (root->step > 50)
		k *= 6;
	else
		if (root->step > 35)
			k *= 2;
	//beg = clock();
	while (1)
	{
		/*
		#pragma omp parrallel sections
		{
		#pragma omp section
		{
		for (int i = 0; i < 10000; i++)
		{
		nextBoard = treePolicy();
		reward = defaultPolicy(nextBoard);
		nextBoard->backup(reward);
		}		pos = action;
		}
		}*/

		for (int i = 0; i < 20000; i++)
		{
			nextBoard = treePolicy();
			reward = defaultPolicy(nextBoard);
			nextBoard->backup(reward);
		}
		break;
		//		if ((clock() - beg) > 25000)
		//		break;
	}

	return root->bestChild(0, 2);

}


void MCTS::release()
{
	root->release();
}


Board* MCTS::treePolicy()
{
	Board *curState = root;

	while (1)
	{
		if (curState->isTerminal())
			break;
		if (curState->needExpand())
			return curState->expand();
		else
			curState = curState->bestChild(C, 1);
	}

	return curState;
}

int MCTS::defaultPolicy(Board* start)
{
	Board* curState = start;
	Board* next;
	while (!(curState->isTerminal()))
	{
		next = curState->randomGo();
		if (curState != start)
		{
			curState->release();
		}
		curState = next;
	}

	int result;
	if (start->player == curState->player)
		result = curState->isWinner();
	else
		result = 1 - curState->isWinner();

	if (curState != start)
	{
		curState->unvisited.clear();
		curState->release();
	}

	return result;
}

